### PR TITLE
Allow user to set keyvault secret client option

### PR DESCRIPTION
**Pull request by zhiyuanliang-ms**
## Why this PR?
#274  

## Visible change

New API
`AzureAppConfigurationKeyVaultOptions.ConfigureClientOptions`

usage:
``` C#
IConfiguration config = new ConfigurationBuilder()
    .AddAzureAppConfiguration(options =>
    {
        options.ConfigureKeyVault(kv =>
        {
            kv.SetCredential(new DefaultAzureCredential());
            kv.ConfigureClientOptions(sco => sco.Retry.MaxAttempts = 3;
        });
    })
    .Build();
```

### CODE REVIEW

File: src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationKeyVaultOptions.cs [hunk 1]
```diff
@@ -31,6 +32,16 @@ public AzureAppConfigurationKeyVaultOptions SetCredential(TokenCredential creden
             return this;
         }
 
+        /// <summary>
+        /// Sets the client options used when connecting to key vaults that have no registered <see cref="SecretClient"/>
+        /// </summary>
+        /// <param name="clientOptions">Default secret client options</param>
```
Comment (amerjusupovic) [inline src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationKeyVaultOptions.cs:38]:
```suggestion
        /// <param name="clientOptions">Default secret client options.</param>
```

File: src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureKeyVaultReference/AzureKeyVaultSecretProvider.cs [hunk 2]
```diff
@@ -115,8 +115,13 @@ private SecretClient GetSecretClient(Uri secretUri)
                 return null;
             }
 
-            client = new SecretClient(new Uri(secretUri.GetLeftPart(UriPartial.Authority)), _keyVaultOptions.Credential);
+            client = new SecretClient(
+                new Uri(secretUri.GetLeftPart(UriPartial.Authority)),
+                _keyVaultOptions.Credential,
+                _keyVaultOptions.ClientOptions ?? new SecretClientOptions());
```
Comment (amerjusupovic) [inline src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureKeyVaultReference/AzureKeyVaultSecretProvider.cs:121]: The `SecretClient` class seems to expect null to be passed for `options`, so we don't need the null check here necessarily. I don't mind it this way, but you could simplify it if you want.

**Comment by zhiyuanliang-ms** [comment]
https://github.com/Azure/AppConfiguration-DotnetProvider/blob/main/src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationOptions.cs#L362

We have `ConfigureClientOptions` and other `ConfigureXXX` methods for `AzureAppConfigurationOptions`. However, this naming convention wasn't followed by `AzureAppConfigurationKeyVaultOptions`. We have `SetXXX` methods for `AzureAppConfigurationKeyVaultOptions`. I think `SetClientOptions` is more appropriate here. Any thought?

@jimmyca15  @amerjusupovic @avanigupta @zhenlan

File: src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureKeyVaultReference/AzureKeyVaultSecretProvider.cs (hunk 2 above)
Comment (zhiyuanliang-ms) [reply src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureKeyVaultReference/AzureKeyVaultSecretProvider.cs:121]: > The SecretClient class seems to expect null to be passed for options

Yea, I know that. 
The reason I add null check here is that I want to make it more robust and don't rely on KeyVault lib to handle the null. 
Besides, if user reads our code, they will know the default option will be used without going to the code base of keyvault lib.

**Comment by amerjusupovic** [comment]
> I think `SetClientOptions` is more appropriate here. What do you think?

I think part of the reason all the methods in `AzureAppConfigurationKeyVaultOptions` start with `Set` is that the methods aren't really configuring anything, it's just passing in a secret resolver, credential, or refresh interval. For example, `AzureAppConfigurationRefreshOptions.SetRefreshInterval` follows that pattern. However, I think the other `Configure` methods in the main `AzureAppConfigurationOptions` involve actually configuring an options class, which is what we're doing in the proposed `SetClientOptions`. 

I think that means `AzureAppConfigurationKeyVaultOptions.ConfigureClientOptions` would make more sense to me. That's just how I see it right now though, any other thoughts?

**Comment by zhenlan** [comment]
It's not just the naming. The benefit of `ConfigureClientOptions` is that users can make necessary client option changes directly inline (#106). They don't have to create a new `SecretClientOptions` object in advance and replace the current one.

```cs
public AzureAppConfigurationOptions ConfigureClientOptions(Action<ConfigurationClientOptions> configure)
```

**Comment by jimmyca15** [comment]
> It's not just the naming. The benefit of `ConfigureClientOptions` is that users can make necessary client option changes directly inline (#106). They don't have to create a new `SecretClientOptions` object in advance and replace the current one.
> 
> ```cs
> public AzureAppConfigurationOptions ConfigureClientOptions(Action<ConfigurationClientOptions> configure)
> ```

I agree here, I prefer the case where the user gets passed a ClientOptions in the callback.

So in your example the usage becomes

```
IConfiguration config = new ConfigurationBuilder()
    .AddAzureAppConfiguration(options =>
    {
        options.ConfigureKeyVault(kv =>
        {
            kv.SetCredential(new DefaultAzureCredential());
            kv.ConfigureClientOptions(sco =>
            {
                        sco.Retry.MaxAttempts = 3;
            });
        });
    })
    .Build();
```

**Comment by zhiyuanliang-ms** [comment]
@jimmyca15 @zhenlan I agree with the proposal. But my only concern is that user will not be able to set SecretClientOptions.Version through the callback pattern.

This is the implementation of `SecretClientOptions`.

``` C#
public class SecretClientOptions : ClientOptions
{
    internal const ServiceVersion LatestVersion = ServiceVersion.V7_5;

    public ServiceVersion Version { get; }

    public bool DisableChallengeResourceVerification { get; set; }

    public SecretClientOptions(ServiceVersion version = ServiceVersion.V7_5)
    {
        Version = version;
        this.ConfigureLogging();
    }
}
```

If user want to configure the keyvault version, it can only be achieved through constructor because `Version` property doesn't have a public setter.

**Comment by jimmyca15** [comment]
@zhiyuanliang-ms I think it's a good call out, but I still think it's the pattern we want.

**Review by jimmyca15 — APPROVED** [review]

**Review by zhenlan — APPROVED** [review]

**Pull request merged**